### PR TITLE
chore: calldata utils improvements & tests

### LINKED ISSUE #1220: utils `calldata` tests coverage
**Is your feature request related to a problem? Please describe.**
- Write tests for `./utils/calldata/*.ts`

**Describe the solution you'd like**
Write tests for each file within `calldata` utils folder

**Describe alternatives you've considered**
_n.a_

**Additional context**
_n.a_

### PR DESCRIPTION
**Pull request by lukasaric**
## Motivation and Resolution

Resolves #1220

Test coverage within `./src/utils/calldata`:

- [x] - `cairo`
- [x] - `validate`
- [x] - `tuple`
- [x] - `byteArray`
- [x] - enum foler ( `CairoResult`, `CairoOption` & `CairoCustomEnum`)
- [x] - parser folder
- [x] - `formatter`

Added ABI factories within mocks

### RPC version (if applicable)

...

## Usage related changes

- Change 1.
- ...

## Development related changes

- Change 1.
- ...

## Checklist:

- [ ] Performed a self-review of the code
- [ ] Rebased to the last commit of the target branch (or merged it into my branch)
- [ ] Linked the issues which this PR resolves
- [ ] Documented the changes in code (API docs will be generated automatically)
- [ ] Updated the tests
- [ ] All tests are passing

### CODE REVIEW

**Review by penovicp — APPROVED** [review]

File: __tests__/utils/calldata/parser/parser.test.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    expect(getAbiVersion([getFunctionAbi('core::bool')])).toEqual(1);
+  });
+
+  test('should return ABI version 0', () => {
+    expect(getAbiVersion([getFunctionAbi('felt')])).toEqual(0);
+  });
+});
+
+describe('isNoConstructorValid', () => {
+  test('should return true if no constructor valid ', () => {
```
Comment (ivpavici) [inline __tests__/utils/calldata/parser/parser.test.ts:37]:
```suggestion
  test('should return true if no constructor valid', () => {
```

File: __tests__/utils/calldata/validate.test.ts [hunk 2]
```diff
@@ -0,0 +1,705 @@
...
+      );
+      expect(result).toBeUndefined();
+    });
+
+    test('should throw an error if parameter is too large', () => {
+      const validateUint = (params: unknown[]) =>
+        validateFields(getFunctionAbi(Uint.u8), params, getAbiStructs(), getAbiEnums());
+
+      const error = new Error(
+        'Validation: Parameter is to large to be typed as Number use (BigInt or String)'
```
Comment (ivpavici) [inline __tests__/utils/calldata/validate.test.ts:175]:
```suggestion
        'Validation: Parameter is too large to be typed as Number, use (BigInt or String)'
```

File: __tests__/utils/calldata/validate.test.ts [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+        getFunctionAbi('struct'),
+        [{ test_name: 'test' }],
+        getAbiStructs(),
+        getAbiEnums()
+      );
+
+      expect(result).toBeUndefined();
+    });
+
+    test('should return void if struct validation passes for Uint 256 or 512', () => {
```
Comment (ivpavici) [inline __tests__/utils/calldata/validate.test.ts:422]: do we miss the 512 case here?

File: __tests__/utils/calldata/validate.test.ts [hunk 4]
```diff
@@ -0,0 +1,705 @@
...
+          [ETH_ADDRESS]: getAbiStructs().struct,
+        };
+
+        validateFields(getFunctionAbi(ETH_ADDRESS), [2n ** 160n], abiStructs, getAbiEnums());
+      }).toThrow(error);
+    });
+
+    test('should throw an error if arg is not an JS object', () => {
+      const error = new Error(
+        'Validate: arg test is cairo type struct (struct), and should be defined as js object (not array)'
```
Comment (ivpavici) [inline __tests__/utils/calldata/validate.test.ts:468]:
```suggestion
        'Validate: arg test is cairo type struct (struct), and should be defined as a js object (not array)'
```

File: __tests__/utils/calldata/validate.test.ts [hunk 5]
```diff
@@ -0,0 +1,705 @@
...
+        getAbiStructs(),
+        abiEnums
+      );
+
+      expect(result).toBeUndefined();
+    });
+
+    test('should throw an error if arg is not an JS object', () => {
+      const error = new Error(
+        'Validate: arg test is cairo type Enum (enum), and should be defined as js object (not array)'
```
Comment (ivpavici) [inline __tests__/utils/calldata/validate.test.ts:536]:
```suggestion
        'Validate: arg test is cairo type Enum (enum), and should be defined as a js object (not array)'
```

File: src/utils/calldata/parser/index.ts [hunk 6]
```diff
@@ -25,7 +25,7 @@ export function isNoConstructorValid(
   method: string,
   argsCalldata: RawArgs,
   abiMethod?: FunctionAbi
-) {
+): boolean {
```
Comment (ivpavici) [inline src/utils/calldata/parser/index.ts:28]: would be ideal to add JSdocs on all these 3 methods as well!

File: src/utils/calldata/validate.ts [hunk 7]
```diff
@@ -320,17 +317,69 @@ const validateNonZero = (parameter: any, input: AbiEntry) => {
...
+ * Flow: Determine type from abi and than validate against parameter
+ *
+ * @param {FunctionAbi} abiMethod - Abi method.
+ * @param {any[]} args - Arguments.
+ * @param {AbiStructs} structs - ABI structs.
+ * @param {AbiEnums} enums - ABI enums.
+ * @returns {void} - Return void if validation passes
+ *
+ * @example
+ *  const functionAbi: FunctionAbi = {
```
Comment (ivpavici) [inline src/utils/calldata/validate.ts:336]:
```suggestion
 * const functionAbi: FunctionAbi = {
```

**Review by ivpavici — APPROVED** [review]

File: __mocks__/factories/abi.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        name: 'test_name',
+        type: 'cairo_struct_variant',
+        offset: 1,
+      },
+    ],
+    size: 2,
+    name: 'test_cairo',
+    type: 'enum',
+  },
+});
```
Comment (tabaktoni) [inline __mocks__/factories/abi.ts:47]: It should be noted what version of ABI this is (For which Cairo versiosns)
Also this should be tested based on latest ABI file, to check if structure is the same.
As such mocks are not refactorable.

File: src/utils/calldata/enum/CairoCustomEnum.ts [hunk 9]
```diff
@@ -45,9 +43,8 @@ export class CairoCustomEnum {
    * @returns the content of the valid variant of a Cairo custom Enum.
    */
   public unwrap(): any {
-    const variants = Object.entries(this.variant);
-    const activeVariant = variants.find((item) => !isUndefined(item[1]));
-    return isUndefined(activeVariant) ? undefined : activeVariant[1];
+    const variants = Object.values(this.variant);
+    return variants.find((item) => !isUndefined(item));
```
Comment (tabaktoni) [inline src/utils/calldata/enum/CairoCustomEnum.ts:47]: Is this same expected case @PhilippeR26

**Review by tabaktoni — CHANGES_REQUESTED** [review]
In general LGTM, just move [abi.ts](https://github.com/starknet-io/starknet.js/pull/1221/files#diff-8d6b0d1bdb45e1458c5c7253f6319e883cf09e5c9e7dd618fe1ace37e9633fcd) back to test, as mocks are non-refatoring element of tests.

File: src/utils/calldata/enum/CairoCustomEnum.ts (hunk 9 above)
Comment (PhilippeR26) [reply src/utils/calldata/enum/CairoCustomEnum.ts:47]: Well,
The current code is not my original code.
The initial intend was to return the content of the only one key that has a content defined, and if all are undefined (a case that should normally not occur) it returns undefined.
I find that ultra compact code is elegant, but is not easy to read!

**Review by tabaktoni — APPROVED** [review]

**Pull request merged**